File: EPIJudge/epi_judge_cpp/lru_cache.cc
#include <vector>
#include <unordered_map>
#include <list>
#include <utility>

#include "test_framework/generic_test.h"
#include "test_framework/test_failure.h"
#include "test_framework/test_utils_serialization_traits.h"

using namespace std;

class LruCache {
 public:
  LruCache(size_t capacity) : cap(capacity) {}

  int Lookup(int isbn) {
    if (d.find(isbn) == d.end()) {
      return -1;
    }

    int price = d[isbn].first;
    q.erase(d[isbn].second);
    q.emplace_front(isbn);
    d[isbn] = {price, q.begin()};
    return price;
  }

  void Insert(int isbn, int price) {
    if (d.find(isbn) != d.end()) {
      Lookup(isbn);
      return;
    }

    if (d.size() == cap) {
      d.erase(q.back());
      q.pop_back();
    }

    q.emplace_front(isbn);
    d[isbn] = {price, q.begin()};
  }

  bool Erase(int isbn) {
    if (d.find(isbn) == d.end()) {
      return false;
    } else {
      q.erase(d[isbn].second);
      d.erase(isbn);
      return true;
    }
  }

 private:
  unordered_map<int, pair<int, list<int>::iterator>> d;
  list<int> q;
  int cap;
};

struct Op {
  std::string code;
  int arg1;
  int arg2;
};

template <>
struct SerializationTraits<Op> : UserSerTraits<Op, std::string, int, int> {};

void RunTest(const std::vector<Op>& commands) {
  if (commands.empty() || commands[0].code != "LruCache") {
    throw std::runtime_error("Expected LruCache as first command");
  }
  LruCache cache(commands[0].arg1);

  for (int i = 1; i < commands.size(); i++) {
    auto& cmd = commands[i];
    if (cmd.code == "lookup") {
      int result = cache.Lookup(cmd.arg1);
      if (result != cmd.arg2) {
        throw TestFailure("Lookup: expected " + std::to_string(cmd.arg2) +
                          ", got " + std::to_string(result));
      }
    } else if (cmd.code == "insert") {
      cache.Insert(cmd.arg1, cmd.arg2);
    } else if (cmd.code == "erase") {
      bool result = cache.Erase(cmd.arg1);
      if (result != cmd.arg2) {
        throw TestFailure("Erase: expected " + std::to_string(cmd.arg2) +
                          ", got " + std::to_string(result));
      }
    } else {
      throw std::runtime_error("Unexpected command " + cmd.code);
    }
  }
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"commands"};
  return GenericTestMain(args, "lru_cache.tsv", &RunTest, DefaultComparator{},
                         param_names);
}

File: EPIJudge/epi_judge_python/lru_cache.py
from sys import exit
from collections import OrderedDict

from test_framework import generic_test, test_utils
from test_framework.test_failure import TestFailure


class LruCache:
    def __init__(self, capacity):
        self.cap = capacity
        self.d = OrderedDict()

    def lookup(self, isbn):
        if isbn in self.d:
            price = self.d.pop(isbn)
            self.d[isbn] = price
            return price
        else:
            return -1

    def insert(self, isbn, price):
        if isbn in self.d:
            self.lookup(isbn)
        else:
            if len(self.d) == self.cap:
                self.d.popitem(last=False)
            self.d[isbn] = price

    def erase(self, isbn):
        if isbn not in self.d:
            return False
        else:
            self.d.pop(isbn)
        return True

def run_test(commands):
    if len(commands) < 1 or commands[0][0] != 'LruCache':
        raise RuntimeError('Expected LruCache as first command')

    cache = LruCache(commands[0][1])

    for cmd in commands[1:]:
        if cmd[0] == 'lookup':
            result = cache.lookup(cmd[1])
            if result != cmd[2]:
                raise TestFailure(
                    'Lookup: expected ' + str(cmd[2]) + ', got ' + str(result))
        elif cmd[0] == 'insert':
            cache.insert(cmd[1], cmd[2])
        elif cmd[0] == 'erase':
            result = 1 if cache.erase(cmd[1]) else 0
            if result != cmd[2]:
                raise TestFailure(
                    'Erase: expected ' + str(cmd[2]) + ', got ' + str(result))
        else:
            raise RuntimeError('Unexpected command ' + cmd[0])


if __name__ == '__main__':
    exit(generic_test.generic_test_main('lru_cache.tsv', run_test))
